hcloud/metadata/client.go

```diff
@@ -18,55 +19,65 @@ const Endpoint = "http://169.254.169.254/hetzner/v1/metadata"
 // Client is a client for the Hetzner Cloud Server Metadata Endpoints.
 type Client struct {
 	endpoint string
+	timeout  time.Duration
 
 	httpClient              *http.Client
 	instrumentationRegistry *prometheus.Registry
 }
 
-// A ClientOption is used to configure a Client.
+// A ClientOption is used to configure a [Client].
 type ClientOption func(*Client)
 
-// WithEndpoint configures a Client to use the specified Metadata API endpoint.
+// WithEndpoint configures a [Client] to use the specified Metadata API endpoint.
 func WithEndpoint(endpoint string) ClientOption {
 	return func(client *Client) {
 		client.endpoint = strings.TrimRight(endpoint, "/")
 	}
 }
 
-// WithHTTPClient configures a Client to perform HTTP requests with httpClient.
+// WithHTTPClient configures a [Client] to perform HTTP requests with httpClient.
 func WithHTTPClient(httpClient *http.Client) ClientOption {
 	return func(client *Client) {
 		client.httpClient = httpClient
 	}
 }
 
-// WithInstrumentation configures a Client to collect metrics about the performed HTTP requests.
+// WithInstrumentation configures a [Client] to collect metrics about the performed HTTP requests.
 func WithInstrumentation(registry *prometheus.Registry) ClientOption {
 	return func(client *Client) {
 		client.instrumentationRegistry = registry
 	}
 }
 
-// NewClient creates a new client.
+// WithTimeout specifies a time limit for requests made by this [Client].
+func WithTimeout(timeout time.Duration) ClientOption {
+	return func(client *Client) {
+		client.timeout = timeout
+	}
+}
+
+// NewClient creates a new [Client] with the options applied.
 func NewClient(options ...ClientOption) *Client {
 	client := &Client{
 		endpoint:   Endpoint,
 		httpClient: &http.Client{},
+		timeout:    1 * time.Second,
```

Comment:
I think 1s as default timeout value is too aggressive, we lose the flexibility of automatically recovering if the network is having some issue, because the request would time out before that.

I would either keep the behavior as it was before, no default timeout, or at least raise it to 5/10/15 seconds.

Reply:
The metadata service is running on each physical host system. There should be no excessive networking between the VM and it, it all happens on the same hardware.

5 seconds sounds good to me :)